src/core/provider/MCAccountValidator.cpp

```diff
@@ -0,0 +1,412 @@
+        mUsername = mEmail;
+    }
+    
+    mProvider = MailProvidersManager::sharedManager()->providerForEmail(mUsername);
+    
+    if (mProvider == NULL) {
+        Array * components;
+        String * domain;
+        
+        components = mUsername->componentsSeparatedByString(MCSTR("@"));
+        if (components->count() >= 2) {
+            domain = (String *) components->lastObject();
+            mResolveMX = new ResolveProviderUsingMXRecord();
+            mResolveMX->setHostname(domain);
+            mQueue->addOperation(mResolveMX);
+        }
+    }
+    else{
+        setup();
+    }
+}
+
+void AccountValidator::cancel()
+{
+    mOperation->cancel();
+    Operation::cancel();
+}
+
+void AccountValidator::operationFinished(Operation * op)
+{
+    opCompleted();
+}
+
+void AccountValidator::setup()
+{
+    if (mResolveMX != NULL) {
+       mProvider = mResolveMX->provider();
+    }
+    
+    if (mProvider != NULL) {
+        mIdentifier = mProvider->identifier();
+        
+        if (mImapServices->count() == 0 and mProvider->imapServices()->count() > 0)
+            mImapServices = mProvider->imapServices();
+        
+        if (mPopServices->count() == 0 and mProvider->popServices()->count() > 0)
+            mPopServices = mProvider->popServices();
+        
+        if (mSmtpServices->count() == 0 and mProvider->smtpServices()->count() > 0)
+            mSmtpServices = mProvider->smtpServices();
+    }
+    
+    if (mImapServices->count() == 0)
+        mImapError = ErrorNoValidServerFound;
+    
+    if (mPopServices->count() == 0)
+        mPopError = ErrorNoValidServerFound;
+    
+    if (mSmtpServices->count() == 0)
+        mSmtpError = ErrorNoValidServerFound;
+    
+    test();
+}
+
+void AccountValidator::opCompleted()
+{
+    ErrorCode error = ErrorNone;
+    
+    if (mCurrentServiceTested == SERVICE_IMAP) {
+        mImapError = ((IMAPOperation *)mOperation)->error();
+        error = mImapError;
+    }
+    else if (mCurrentServiceTested == SERVICE_POP) {
+        mPopError = ((POPOperation *)mOperation)->error();
+        error = mPopError;
+    }
+    else if (mCurrentServiceTested == SERVICE_SMTP) {
+        mSmtpError = ((SMTPOperation *)mOperation)->error();
+        error = mSmtpError;
+    }
+    
+    if (error == ErrorNone) {
+        mCurrentServiceTested ++;
+    }
+    else {
+        mCurrentServiceIndex ++;
+    }
+    
+    test();
+}
+
+/**
+ Each service(IMAP/POP/SMTP) is tested one after the other.
+ For each service we test each server details (NetService),
+ Until either:  
+    we find on that works and returns ErrorNone in opCompleted().
+    we have gone trough the Array of NetService for that service and test() is recalled for the next service.
+ */
+void AccountValidator::test()
+{
+    if (mCurrentServiceTested == SERVICE_IMAP) {
+        if (mCurrentServiceIndex < mImapServices->count()) {
+            IMAPAsyncSession *imapSession = new IMAPAsyncSession();
```

Comment:
I've added the sessions as properties so we can keep the IMAP session and properly release the sessions others at the end of the operation. Or release them all if canceled.
Do you want to keep the session to maybe make it accessible after the operation?